Avoid failures when redirecting from incomplete/invalid Delta Lake tables

nineinchnick opened the pull request:
<!-- Thank you for submitting a pull request! Find more information 
at https://trino.io/development/process.html, 
at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md 
and contact us on #core-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description

Don't attempt to read Delta Lake table metadata when performing table redirect to a different catalog. This allows redirecting from incomplete or invalid Delta Lake tables.

Without the fix, queries fail with:
```
io.trino.testing.QueryFailedException: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Error: type expected at the position 0 of 'timestamp_ntz' but 'timestamp_ntz' is found.

	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:134)
	at io.trino.testing.DistributedQueryRunner.executeInternal(DistributedQueryRunner.java:565)
	at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:548)
	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:313)
	at io.trino.testing.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:308)
	at io.trino.plugin.deltalake.TestDeltaLakeSharedGlueMetastoreWithTableRedirections.testUnsupportedHiveTypeRedirect(TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:180)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1458)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2034)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:189)
	Suppressed: java.lang.Exception: SQL: SHOW CREATE TABLE hive_with_redirections.test_shared_schema_4jlq4sptnq.unsupported_types
		at io.trino.testing.DistributedQueryRunner.executeInternal(DistributedQueryRunner.java:572)
		... 9 more
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Error: type expected at the position 0 of 'timestamp_ntz' but 'timestamp_ntz' is found.
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2085)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4017)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4040)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4989)
	at io.trino.cache.EvictableCache.get(EvictableCache.java:147)
	at com.google.common.cache.AbstractLoadingCache.getUnchecked(AbstractLoadingCache.java:53)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.getOptional(CachingHiveMetastore.java:285)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.getTable(CachingHiveMetastore.java:431)
	at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.getTable(SemiTransactionalHiveMetastore.java:276)
	at io.trino.plugin.hive.HiveMetadata.getView(HiveMetadata.java:2899)
	at io.trino.spi.connector.ConnectorMetadata.isView(ConnectorMetadata.java:1005)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.isView(ClassLoaderSafeConnectorMetadata.java:721)
	at io.trino.tracing.TracingConnectorMetadata.isView(TracingConnectorMetadata.java:850)
	at io.trino.metadata.MetadataManager.lambda$isView$37(MetadataManager.java:1499)
	at java.base/java.util.Optional.map(Optional.java:260)
	at io.trino.metadata.MetadataManager.isView(MetadataManager.java:1495)
	at io.trino.tracing.TracingMetadata.isView(TracingMetadata.java:886)
	at io.trino.sql.rewrite.ShowQueriesRewrite$Visitor.showCreateTable(ShowQueriesRewrite.java:616)
	at io.trino.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowCreate(ShowQueriesRewrite.java:519)
	at io.trino.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowCreate(ShowQueriesRewrite.java:219)
	at io.trino.sql.tree.ShowCreate.accept(ShowCreate.java:60)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.rewrite.ShowQueriesRewrite.rewrite(ShowQueriesRewrite.java:216)
	at io.trino.sql.rewrite.StatementRewrite.rewrite(StatementRewrite.java:54)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:93)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:87)
	at io.trino.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:289)
	at io.trino.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:222)
	at io.trino.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:892)
	at io.trino.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:153)
	at io.trino.$gen.Trino_testversion____20241010_143532_71.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Error: type expected at the position 0 of 'timestamp_ntz' but 'timestamp_ntz' is found.
	at io.trino.cache.EmptyCache.get(EmptyCache.java:100)
	at io.trino.cache.EmptyCache.get(EmptyCache.java:58)
	at com.google.common.cache.AbstractLoadingCache.getUnchecked(AbstractLoadingCache.java:53)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.getOptional(CachingHiveMetastore.java:285)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.getTable(CachingHiveMetastore.java:431)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:436)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:169)
	at io.trino.cache.EvictableCache$TokenCacheLoader.load(EvictableCache.java:471)
	at io.trino.cache.EvictableCache$TokenCacheLoader.load(EvictableCache.java:457)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3574)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2316)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2189)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2079)
	... 36 more
Caused by: java.lang.IllegalArgumentException: Error: type expected at the position 0 of 'timestamp_ntz' but 'timestamp_ntz' is found.
	at io.trino.metastore.type.TypeInfoUtils$TypeInfoParser.expect(TypeInfoUtils.java:193)
	at io.trino.metastore.type.TypeInfoUtils$TypeInfoParser.expect(TypeInfoUtils.java:176)
	at io.trino.metastore.type.TypeInfoUtils$TypeInfoParser.parseType(TypeInfoUtils.java:232)
	at io.trino.metastore.type.TypeInfoUtils$TypeInfoParser.parseTypeInfos(TypeInfoUtils.java:154)
	at io.trino.metastore.type.TypeInfoUtils.getTypeInfosFromTypeString(TypeInfoUtils.java:357)
	at io.trino.metastore.type.TypeInfoUtils.getTypeInfoFromTypeString(TypeInfoUtils.java:362)
	at io.trino.metastore.HiveType.valueOf(HiveType.java:116)
	at io.trino.plugin.hive.metastore.glue.GlueConverter.fromGlueColumn(GlueConverter.java:284)
	at io.trino.plugin.hive.metastore.glue.GlueConverter.lambda$fromGlueColumns$1(GlueConverter.java:272)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1709)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:727)
	at io.trino.plugin.hive.metastore.glue.GlueConverter.fromGlueColumns(GlueConverter.java:273)
	at io.trino.plugin.hive.metastore.glue.GlueConverter.fromGlueTable(GlueConverter.java:188)
	at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getTableInternal(GlueHiveMetastore.java:495)
	at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.lambda$getTable$18(GlueHiveMetastore.java:485)
	at io.trino.plugin.hive.metastore.glue.NoopGlueCache.getTable(NoopGlueCache.java:66)
	at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.getTable(GlueHiveMetastore.java:485)
	at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:436)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:169)
	at io.trino.cache.EmptyCache.lambda$get$0(EmptyCache.java:58)
	at io.trino.cache.EmptyCache.get(EmptyCache.java:94)
	... 48 more

```

<!-- Provide details that help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

pajaks commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +119,58 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+            glueClient.createTable(createTableRequest);
+
+            // The table is incomplete, so it is expected to fail after reading the metadata from Glue
+            assertQueryFails("SHOW CREATE TABLE hive_with_redirections." + schema + "." + tableName, "Metadata not found in transaction log for " + schema + "." + tableName);
```
Can you try to prepare table, so that query succeeds?

nineinchnick replied:
Done!

findinpath commented:
Please consider adding a stacktrace in the description of the bugfix to describe better your solution to the problem.

nineinchnick commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -211,6 +213,7 @@ private GlueHiveMetastore(
             int partitionSegments,
             boolean assumeCanonicalPartitionKeys,
             Set<TableKind> visibleTableKinds,
+            boolean ignoreIncompleteDeltaTables,
```
Removed

nineinchnick commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -139,6 +139,11 @@ public static DatabaseInput toGlueDatabaseInput(Database database)
+        return fromGlueTable(glueTable, databaseName, false);
+    }
+
+    public static Table fromGlueTable(software.amazon.awssdk.services.glue.model.Table glueTable, String databaseName, boolean ignoreIncompleteDeltaTables)
```
Renamed to `ignoreDeltaLakeColumns`

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -20,9 +20,18 @@
 import io.trino.testing.DistributedQueryRunner;
```
Shortened

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -139,6 +139,11 @@ public static DatabaseInput toGlueDatabaseInput(Database database)
+        return fromGlueTable(glueTable, databaseName, false);
+    }
+
+    public static Table fromGlueTable(software.amazon.awssdk.services.glue.model.Table glueTable, String databaseName, boolean ignoreDeltaLakeColumns)
```
`ignoreDeltaLakeColumns` -> `ignoreColumns`

We may be dealing similarly with an Iceberg or Hudi table

nineinchnick replied:
I can't write a test for other connectors.

findinpath replied:
Let's drop `boolean ignoreDeltaLakeColumns` parameter.
I think we can agree that if `isDeltaLakeTable(tableParameters)` should be sufficient to allow us to ignore the columns.

FWIW I'm ok with approving the PR with this requested change.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -176,7 +181,12 @@ else if (isIcebergTable(tableParameters)) {
         else {
             boolean isCsv = sd.serdeInfo() != null && HiveStorageFormat.CSV.getSerde().equals(sd.serdeInfo().serializationLibrary());
-            dataColumns = fromGlueColumns(sd.columns(), ColumnType.DATA, isCsv);
+            if (isDeltaLakeTable(tableParameters) && ignoreDeltaLakeColumns) {
```
We should always ignore columns for 

```
if (isIcebergTable(table) || isDeltaLakeTable(table))
```

I'm not sure whether `ignoreDeltaLakeColumns` check is actually relevant.

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,75 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+    @Test
+    public void testUnsupportedHiveTypeRedirect()
+    {
+        String tableName = "unsupported_types";
```
The whole schema has a random name: https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeSharedMetastoreWithTableRedirectionsTest.java#L28

I figured we don't need random table names, since this is not running in parallel. Also, the existing tests in this class don't use random table names.

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,75 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+        String tableName = "unsupported_types";
+        // use another complete table location so `SHOW CREATE TABLE` doesn't fail on reading metadata
+        String location;
+        try (GlueClient glueClient = GlueClient.create()) {
```
This is about tables created by Spark. I based this on a report I received, and I don't have a better reproduction - that is, I don't know _exactly_ how Spark creates this table.

I'll try to explain this in a comment.

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,75 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+                        "spark.sql.sources.provider", "delta",
+                        "delta.lastCommitTimestamp", "1720682605000",
+                        "spark.sql.create.version", "3.4.1",
+                        "delta.lastUpdateVersion", "0"))
```
I removed most parameters

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,75 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+        try (GlueClient glueClient = GlueClient.create()) {
+            glueClient.createTable(createTableRequest);
+
+            // The table is incomplete, so it is expected to fail after reading the metadata from Glue
```
Sorry, this is a left-over comment from before I addressed https://github.com/trinodb/trino/pull/23730#discussion_r1793478351, I removed it.

nineinchnick commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,67 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+                    .build());
+            location = existingTable.table().storageDescriptor().location();
+        }
+        // Create a table directly in Glue, simulating an external table being created in Spark,
```
@ebyhr PTAL if this is a good enough explanation

anusudarsan replied:
maybe extract a method so you dont need the comment

findinpath replied:
`timestamp_ntz` is actually supported for Trino Delta Lake connector.
The issue we're dealing with here is that  `GlueConverter` knows only about the types mapped under `io.trino.metastore.HiveType`.
We could consider using a different type mapping strategy based on the table type (hive/delta/iceberg).
However, in case of Delta (and Iceberg) we have the luxury to disregard completely what the metastore is saying about the columns because this information is retrieved from the metadata of the table from the object storage (and not the metastore).

nineinchnick replied:
Reworded the comment, PTAL

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -176,7 +181,12 @@ else if (isIcebergTable(tableParameters)) {
         else {
             boolean isCsv = sd.serdeInfo() != null && HiveStorageFormat.CSV.getSerde().equals(sd.serdeInfo().serializationLibrary());
-            dataColumns = fromGlueColumns(sd.columns(), ColumnType.DATA, isCsv);
+            if (isDeltaLakeTable(tableParameters) && ignoreDeltaLakeColumns) {
```
Similarly to Iceberg handling, consider placing this in a block 

```
else if (isDeltaLakeTable(tableParameters)) {
....
}
```

The `partitionColumns` are not stored AFAIK in the metastore.

Check also whether the same storage as for the Iceberg tables makes sense in this case.
```
storage = FAKE_PARQUET_STORAGE;
```

I'm inclined to advocate for setting always for Iceberg & Delta Lake tables:

```
dataColumns = ImmutableList.of(FAKE_COLUMN);
```

We don't use nowhere this parsed information. The downside of trying to parse this information is that sometimes we may get burned (as in the situation you're trying to fix).

nineinchnick replied:
> Similarly to Iceberg handling, consider placing this in a block

Why? Can you add a code suggestion?

> Check also whether the same storage as for the Iceberg tables makes sense in this case.

Isn't this out of the scope of this PR?

> I'm inclined to advocate for setting always for Iceberg & Delta Lake tables

I'd agree, but I don't feel confident expanding the scope of this for Iceberg, as I can't write a new test for Iceberg, and I don't know if the existing tests cover that.

findinpath reviewed (commented):
Almost there. :)

findinpath reviewed (approved):
LGTM % minor change in the test comment wording https://github.com/trinodb/trino/pull/23730/files#r1806610589

wendigo commented:
Test failure seems related

ebyhr commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -176,7 +176,12 @@ else if (isIcebergTable(tableParameters)) {
         else {
             boolean isCsv = sd.serdeInfo() != null && HiveStorageFormat.CSV.getSerde().equals(sd.serdeInfo().serializationLibrary());
-            dataColumns = fromGlueColumns(sd.columns(), ColumnType.DATA, isCsv);
+            if (isDeltaLakeTable(tableParameters)) {
```
Any reason not to update the above condition handling Iceberg format at L166? 
```java
        else if (isIcebergTable(tableParameters)) {
```

nineinchnick replied:
Are you asking if the condition on `isDeltaLakeTable(tableParameters)` can be moved one lever higher, or can it be added to the condition for Iceberg? I'm not sure where `storage` is used, so I wanted to preserve how it's built. `dataColumns` for Iceberg is also built differently. I can move the condition one level higher, but that'll duplicate some code. LMK if that's ok.

ebyhr replied:
I meant adding to the Iceberg condition, but a separate condition seems better. Can we change like this? 
```java
        else if (isIcebergTable(tableParameters)) {
        ...
        }
        else if (isDeltaLakeTable(tableParameters)) {
            dataColumns = ImmutableList.of(FAKE_COLUMN);
            partitionColumns = ImmutableList.of();
            storage = fromGlueStorage(sd, databaseName + "." + glueTable.name());
        }
```

By the way, I think the existing logic still fails when `timestamp_ntz` type is used in partition columns. Could you add such a partition column in a test?

nineinchnick replied:
Much better, and great catch, thanks!

ebyhr commented on plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeSharedGlueMetastoreWithTableRedirections.java:
```diff
@@ -108,4 +122,67 @@ protected String getExpectedDeltaLakeCreateSchema(String catalogName)
+        try (GlueClient glueClient = GlueClient.create()) {
+            glueClient.createTable(createTableRequest);
+
+            String tableDefinition = (String) computeActual("SHOW CREATE TABLE hive_with_redirections." + schema + "." + tableName).getOnlyValue();
```
nit: We could use `computeScalar` instead.

ebyhr commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestGlueConverter.java:
```diff
@@ -403,11 +402,7 @@ void testDeltaTableNonNullStorageDescriptor()
-                .collect(toImmutableSet())).isEqualTo(glueTable.storageDescriptor().columns().stream()
-                .map(software.amazon.awssdk.services.glue.model.Column::name)
-                .collect(toImmutableSet()));
+        assertThat(trinoTable.getDataColumns().size()).isEqualTo(1);
```
AssertJ provides `hasSize` method.

nineinchnick replied:
Done, I also fixed other occurrences in a separate commit.

ebyhr commented:
@nineinchnick Could you confirm CI failure? It looks related to this change.

nineinchnick commented:
Fixed, sorry about that. I thought we could increase the coverage in tests with the timestamp column, but it's not relevant to this PR anyway, so I reverted that.

ebyhr commented:
/test-with-secrets sha=bd98c080ad5c0510b0cd2ad4f8b010bc64bd752e

pajaks commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueConverter.java:
```diff
@@ -174,6 +174,11 @@ else if (isIcebergTable(tableParameters)) {
             partitionColumns = ImmutableList.of();
             storage = FAKE_PARQUET_STORAGE;
         }
+        else if (isDeltaLakeTable(tableParameters)) {
```
Should similar fix be done for v1 version `GlueToTrinoConverter` ?

nineinchnick replied:
I'm not sure

ebyhr replied:
It's already handled in my understanding: https://github.com/trinodb/trino/blob/e3b62bd026f3941ffdb19baee8f88aa030f4cb1a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/v1/converter/GlueToTrinoConverter.java#L156

nineinchnick replied:
I applied the same fix for v1. I ran the test with v1 manually, but did not add it in the commit, since v1 is to be removed soon.

@pajaks thanks for spotting this!

ebyhr commented:
/test-with-secrets sha=388c35db830c95fdd0b92b0b89fdee3897228e61

pajaks commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/v1/converter/GlueToTrinoConverter.java:
```diff
@@ -156,11 +156,24 @@ public static Table convertTable(com.amazonaws.services.glue.model.Table glueTab
+        else if (isDeltaLakeTable(tableParameters)) {
+            tableBuilder.setDataColumns(ImmutableList.of(new Column("dummy", HIVE_INT, Optional.empty(), ImmutableMap.of())));
+            tableBuilder.setPartitionColumns(ImmutableList.of());
+            if (storageDescriptor.isEmpty()) {
```
This should will be handled above in `(storageDescriptor.isEmpty() && isDeltaLakeTable(tableParameters))`, right?
You could remove it from here, or remove from if condition

nineinchnick replied:
I removed it from the previous condition, thanks for spotting this!

ebyhr commented:
/test-with-secrets sha=7d601433216c22ae08361dc4a50dafcc07e71dc0

ebyhr commented:
/test-with-secrets sha=7d601433216c22ae08361dc4a50dafcc07e71dc0

nineinchnick commented:
@ebyhr you should try f75570104e053d68da4f1de742f45cc880a02920, not 7d60143, but it passed in https://github.com/trinodb/trino/pull/23974 anyway
